Merge Explorer & Everything Plugins

taooceros opened the pull request:
#1502

1. Shared Interface for Windows Index and Everything Index
2. Settings Merge (Part 1)
3. Include Everything dll

TODO:
- [x] finish Path traversal unify
- [x] implement Engine selection
- [x] #526 
- [x] #646 
- [x] implement Everything auto download
- [x] Context Menu for windows (ported from everything)
- [x] Everything Translation

taooceros commented:
Thinking about enumeration design. Probably it would be better to only use index search when doing recursive search for enumerating a directory, and do a direct search when only traverse the first level. @jjw24 what do you think

jjw24 commented:
I think directory recursive search with Windows Search could be restricted by user's indexing options/excluded directories, so allowing an option to use DirectoryInfo search method is preferred per current design in this area.

What do you mean by direct search when only traverse the first leve?

taooceros commented:
> I think directory recursive search with Windows Search could be restricted by user's indexing options/excluded directories, so allowing an option to use DirectoryInfo search method is preferred per current design in this area.
> 
> What do you mean by direct search when only traverse the first leve?

oh I mean we may always use the DirectoryInfo Search for the non-recursive traversal.

taooceros commented:
@onesounds mind creating a engine selection ui for this? I am not sure whether it is a good practice to use interface for purpose selection, but probably we can adjust that easier.

- IndexSearchEngine
  - WindowsIndex
  - Everything
- ContentSearchEngine
  - WindowsIndex
  - Everything
- PathEnumerator
  - Direct
  - WindowsIndex
  - Everything

taooceros commented:
@Flow-Launcher/team I have almost finished most features and the UI Refactor. There're some small features in everything plugin that I haven't ported, but it will be fairly easy to do so. Anyone want to give this pr a check?

taooceros commented:
@jjw24 feel free to review. The remaining part is only about everything's translation, and I will try to fix it through crowdin recently.

jjw24 commented:
Could you fix the build failure please :)

jjw24 commented:
I haven't had a chance to look at the code yet but just thought of something- have you added handling of situation where user already has Everything service/plugin installed?

taooceros commented:
> I haven't had a chance to look at the code yet but just thought of something- have you added handling of situation where user already has Everything service/plugin installed?

not for everything plugin. For everything service, I just copied the code from everything plugin, so as long as it is handled there, we are fine.

taooceros commented:
I will fix the conflict tmr.

jjw24 commented:
@taooceros could you please fix the CI

jjw24 commented:
Couple of things:
1. Path search seems to be always using index search, 'Use Index Search For Path Search' unticked and selecting 'Direct Enumeration' do nothing.
2. Please complete Everything's translation, the keys should at least exist.
3. Selecting Everything should prompt user if they want to download and install Everything, when 'Yes' flow should automatically download and install it.
4. Clicking on the plugin's settings:
![image](https://user-images.githubusercontent.com/26427004/201588745-4b0b8eff-eb0d-4547-b1bb-126cc103ba8b.png)

5. What is 'Launch Hidden' and I remember it's part of Everything plugin? Should move to the Everything setting tab if the case.
6. Using 'Index Search Excluded Path' is not working, could be same issue as point 1.

jjw24 commented:
@taooceros please update description on what you have tested.

onesounds commented:
Fix https://github.com/Flow-Launcher/Flow.Launcher/issues/1495

taooceros commented:
> Couple of things:
> 
>     1. Path search seems to be always using index search, 'Use Index Search For Path Search' unticked and selecting 'Direct Enumeration' do nothing.

It works, but the path equality check have some issue due to previous change to make subtitle only display parent path, which makes everything duplicate except the first one. fixed. 


>     2. Please complete Everything's translation, the keys should at least exist.

Fixed


>     3. Selecting Everything should prompt user if they want to download and install Everything, when 'Yes' flow should automatically download and install it.
>

I think we can do it when user actually search (when the result display everything not working).

>     4. Clicking on the plugin's settings:
>        ![image](https://user-images.githubusercontent.com/26427004/201588745-4b0b8eff-eb0d-4547-b1bb-126cc103ba8b.png)
> 

I cannot reproduce

>     5. What is 'Launch Hidden' and I remember it's part of Everything plugin? Should move to the Everything setting tab if the case.

This is a weird setting and I think we have made the decision to remove it.

>     6. Using 'Index Search Excluded Path' is not working, could be same issue as point 1.

Will handle that soon.

jjw24 commented:
Translations are done?

jjw24 commented on Plugins/Flow.Launcher.Plugin.Explorer/Languages/en.xaml:
```diff
@@ -17,13 +17,21 @@
     <system:String x:Key="plugin_explorer_indexsearchexcludedpaths_header">Index Search Excluded Paths</system:String>
+    <system:String x:Key="plugin_explorer_use_location_as_working_dir">Use search result's location as executable working directory</system:String>
     <system:String x:Key="plugin_explorer_usewindowsindexfordirectorysearch">Use Index Search For Path Search</system:String>
     <system:String x:Key="plugin_explorer_usewindowsindexfordirectorysearch_tooltip">Turning this on will return indexed directories/files faster, but if a directory/file is not indexed it will not show up. If a directory/file has been added to Index Search Excluded Path then it will still show up even if this option is on</system:String>
```
This should change to `If directory is indexed, turning this on will return indexed directories/files faster, but if a directory/file is not indexed it will still use path enumeration so it can still be shown.`
Because UseWindowsIndexForDirectorySearch call checks and returns the outcome of whether the path is indexed.

jjw24 replied:
hmm on second thought this option seems confusing now because feels like a 'maybe use index search' but also depends also on Paths Enumeration Engine.

jjw24 commented on Plugins/Flow.Launcher.Plugin.Explorer/Settings.cs:
```diff
@@ -38,8 +56,94 @@ public class Settings
+            [Description("plugin_explorer_engine_everything")]
+            Everything,
+            [Description("plugin_explorer_path_enumeration_engine_none")]
+            Direct
```
Maybe call this DirectEnumeration so it's consistent with UI

jjw24 commented on Plugins/Flow.Launcher.Plugin.Explorer/Settings.cs:
```diff
@@ -1,22 +1,40 @@
+        public ObservableCollection<AccessLink> QuickAccessLinks { get; set; } = new();
 
         // as at v1.7.0 this is to maintain backwards compatibility, need to be removed afterwards.
-        public List<AccessLink> QuickFolderAccessLinks { get; set; } = new List<AccessLink>();
+        public ObservableCollection<AccessLink> QuickFolderAccessLinks { get; set; } = new();
```
Remove this and the other call, no longer needed.

jjw24 reviewed (commented):
>3. Selecting Everything should prompt user if they want to download and install Everything, when 'Yes' flow should automatically download and install it.

This is still not fixed. Using Everything engine for index search/content search without everything installed just errors out:
![image](https://user-images.githubusercontent.com/26427004/204387761-67138494-e7e6-4459-b1a9-f541c3b6c6fa.png)

jjw24 commented:
>5. What is 'Launch Hidden' and I remember it's part of Everything plugin? Should move to the Everything setting tab if the case.
This is a weird setting and I think we have made the decision to remove it.

Please remove it.

jjw24 commented:
Shouldn't these settings be moved to Everything settings tab?

![image](https://user-images.githubusercontent.com/26427004/204389740-a3ecdd8c-2db6-4320-9163-0166a2d169c9.png)

taooceros commented:
will check that later today

jjw24 commented:
Could this be Everything is installed but not running?

taooceros commented:
> Could this be Everything is installed but not running?

I do haven't uninstalled everything, but I believe both should throw IPCError? Not so sure why it may be different. Will check tongiht

jjw24 commented:
Changing action keywords do not take effect on first try. Repro:
Remove the settings.json file in user data directory. Start up flow, go to Explorer and change Path action keyword. Trigger the keyword and it does not happen. Only after changing the action keyword again then it will trigger.

jjw24 commented:
Please address:
1. When Everything is selected as Index Search Engine, press and hold any key will cause below error:
![image](https://user-images.githubusercontent.com/26427004/204681320-20cf5d5d-bcde-4626-938a-1ead336da8d1.png)

2. Bring this message back when Windows Index is disabled:
![image](https://user-images.githubusercontent.com/26427004/204681580-1233e8e8-321c-4096-9fc8-d040f159ea41.png)

3. Hide the search window when the prompt to install Everything is shown:
![image](https://user-images.githubusercontent.com/26427004/204682662-bedd053d-8fb2-4e7b-8855-1a1512304822.png)

4. Please add icon for this result:
![image](https://user-images.githubusercontent.com/26427004/204683359-1565ec9b-4ea6-4a7b-b291-47b8b2eee4ef.png)

jjw24 commented:
> Please address:
> 
> 1. When Everything is selected as Index Search Engine, press and hold any key will cause below error:
>    ![image](https://user-images.githubusercontent.com/26427004/204681320-20cf5d5d-bcde-4626-938a-1ead336da8d1.png)
> 2. Bring this message back when Windows Index is disabled:
>    ![image](https://user-images.githubusercontent.com/26427004/204681580-1233e8e8-321c-4096-9fc8-d040f159ea41.png)
> 3. Hide the search window when the prompt to install Everything is shown:
>    ![image](https://user-images.githubusercontent.com/26427004/204682662-bedd053d-8fb2-4e7b-8855-1a1512304822.png)
> 4. Please add icon for this result:
>    ![image](https://user-images.githubusercontent.com/26427004/204683359-1565ec9b-4ea6-4a7b-b291-47b8b2eee4ef.png)

@taooceros I am fixing 2,3,4

taooceros commented:
the concurrency bug is ....very weird but interesting (I know what's causing it, and try to find a good solution).
